design/src/rfcs/rfc0031_providing_fallback_credentials_on_timeout.md

```diff
@@ -0,0 +1,251 @@
+
+Referring back to the code snippet above, we analyze two relevant cases (and suppose provider 2 below must meet REQ 1 and REQ 2 in each case):
+
+**Case 1:** Provider 2 successfully loaded credentials but later failed to do so because an external timeout kicked in.
+
+<p align="center">
+<img width="750" alt="chain-provider-ext-timeout-1" src="https://user-images.githubusercontent.com/15333866/212421638-d08e4821-2dbe-497f-82c5-c78aab8acbe9.png">
+</p>
+
+The figure above illustrates an example. This `CredentialsProviderChain` consists of three credentials providers. When `CredentialsProviderChain::provide_credentials` is called, provider 1's `provide_credentials` is called but does not find credentials so passes the torch to provider 2, which in turn successfully loads credentials and returns them. The next time the method is called, provider 1 does not find credentials but neither does provider 2 this time, because an external timeout by `timeout_future` given to the whole chain kicked in and the future is dropped while provider 2's `provide_credentials` was running. Given the functional requirements, provider 2 should return the previously available credentials but today the code snippet from `LazyCredentialsCache` returns a `CredentialsError::ProviderTimedOut` instead.
+
+**Case 2:** Provider 2 successfully loaded credentials but later was not reached because its preceding provider was still running when an external timeout kicked in.
+
+<p align="center">
+<img width="750" alt="chain-provider-ext-timeout-2" src="https://user-images.githubusercontent.com/15333866/212421712-8c6eab11-a0c1-4229-8ba3-67b0bb6056e7.png">
+</p>
+
+The figure above illustrates an example with the same setting as the previous figure. Again, when `CredentialsProviderChain::provide_credentials` is called the first time, provider 1 does not find credentials but provider 2 does. The next time the method is called, provider 1 is still executing `provide_credentials` and then an external time by `timeout_future` kicked in. Consequently, the execution of `CredentialsProviderChain::provide_credentials` has been terminated. Given the functional requirements, provider 2 should return the previously available credentials but today the code snippet from `LazyCredentialsCache` returns `CredentialsError::ProviderTimedOut` instead.
+
+
+Proposal
+--------
+To address the problem in the previous section, we propose to add a new method to the `ProvideCredentials` trait called `on_timeout` that looks something like this:
+```rust
+mod future {
+    // --snip--
+
+    pub struct OnTimeout<'a>(NowOrLater<Option<Credentials>, BoxFuture<'a, Option<Credentials>>>);
+
+    // impls for OnTimeout similar to those for the ProvideCredentials future newtype
+}
+
+pub trait ProvideCredentials: Send + Sync + std::fmt::Debug {
+    // --snip--
+
+    fn on_timeout<'a>(&'a self) -> future::OnTimeout<'a> {
```

Comment:
Is there any reason this shouldn't be synchronous? What's the use-case for asynchronously retrieving the credentials on timeout? Would we have a timeout for the `on_timeout`?

Reply:
No reason it shouldn't. I think each has pros/cons:
- synchronous
**pros**: Easier to reason about. 
**cons**: Will be a breaking change if we want to switch to async in the future.
- asynchronous
**pros**: It's async from the beginning, so less likely to introduce a breaking change.
**cons**: Exactly what you brought up, may need to specify a timeout.

Currently, there are no concrete use cases requiring this to be async.

Reply:
An async primitive is more suitable as discussed (since we do not have an async critical section). The method is a fallback plan so the execution should not take too long. d413e58 mimics `FutureExt::now_or_never` using our `NowOrLater` with `OnlyReady`, which means no need to have yet another timeout for the method.

Reply:
> Would we have a timeout for the on_timeout?

It is possible, although I don't think we specify that as an input argument to `fallback_on_interrupt` for the same reason I described in the Alternative section in the RFC. Whoever uses `fallback_on_timeout` can race it against a timeout future just like `LazyCredentialsCache` does with `provide_credentials` today. It just that the timeout for `fallback_on_timeout` will probably way shorter than that for `provide_credentials`, as the execution of `fallback_on_timeout` shouldn't take too long.

Haven't decided whether the timeout for `fallback_on_timeout` should be introduced today (something like `load_timeout` for `LazyCredentialsCache`). If we are to introduce it later, we need to see if we can do so without causing breaking changes.

Reply:
> Would we have a timeout for the on_timeout?

It is possible, although I don't think we specify that as an input argument to `fallback_on_interrupt` for the same reason I described in the Alternative section in the RFC. Whoever uses `fallback_on_timeout` can race it against a timeout future just like `LazyCredentialsCache` does with `provide_credentials` today. It just that the timeout for `fallback_on_timeout` will probably be way shorter than that for `provide_credentials`, as the execution of `fallback_on_timeout` shouldn't take too long.

Haven't decided whether the timeout for `fallback_on_timeout` should be introduced today (something like `load_timeout` for `LazyCredentialsCache`). If we are to introduce it later, we need to see if we can do so without causing breaking changes.